tools/dds/dds-server/lrs-device-controller.cpp

```diff
@@ -2,99 +2,205 @@
+                                              && sp.stream_index() == stream_name_to_index( stream_name )
+                                              && sp.fps() == profile->frequency()
+                                              && sp.format() == profile->format().to_rs2()
+                                              && video_params_match;
+                                      } );
+    if ( profile_iter == sensor_stream_profiles.end() )
+    {
+        throw std::runtime_error( "Could not find required profile" );
+    }
+
+    return *profile_iter;
+}
+
+lrs_device_controller::lrs_device_controller( rs2::device dev, std::shared_ptr< realdds::dds_device_server > dds_device_server )
     : _rs_dev( dev )
+    , _dds_device_server( dds_device_server )
 {
+    if ( ! _dds_device_server )
+        throw std::runtime_error( "Empty dds_device_server" );
+
+    _dds_device_server->on_open_streams( [&]( const json & msg ) { start_streaming( msg ); } );
+    _dds_device_server->on_close_streams( [&]( const json & msg ) { stop_streaming( msg ); } );
+
+    //query_sensors returns a copy of the sensors, we keep it to use same copy throughout the run time.
+    //Otherwise problems could arise like opening streams and they would close at start_streaming scope end.
+    _sensors = _rs_dev.query_sensors();
+
     _device_sn = _rs_dev.get_info( RS2_CAMERA_INFO_SERIAL_NUMBER );
     LOG_DEBUG( "LRS device manager for device: " << _device_sn << " created" );
 }
 
 
 lrs_device_controller::~lrs_device_controller()
 {
-    stop_all_streams();
     LOG_DEBUG( "LRS device manager for device: " << _device_sn << " deleted" );
 }
-void lrs_device_controller::start_stream( rs2::stream_profile sp, frame_callback_type cb )
+
+void lrs_device_controller::start_streaming( const json & msg )
 {
-    switch( sp.stream_type() )
+    auto msg_profiles = msg["stream-profiles"];
+
+    std::vector< rs2::stream_profile > rs_profiles_to_open;
+    std::vector< std::pair < std::string, image_header > > realdds_streams_to_start;
+
+    size_t sensor_index = 0;
+    for ( auto it = msg_profiles.begin(); it != msg_profiles.end(); ++it )
     {
-    case RS2_STREAM_COLOR: {
-        auto cs = _rs_dev.first< rs2::color_sensor >();
-        stream_to_rs2_sensor[RS2_STREAM_COLOR] = std::make_shared< lrs_sensor_streamer >( cs, sp, cb );
-    }
-    break;
-    case RS2_STREAM_DEPTH: {
-        auto ds = _rs_dev.first< rs2::depth_sensor >();
-        stream_to_rs2_sensor[RS2_STREAM_DEPTH] = std::make_shared< lrs_sensor_streamer >( ds, sp, cb );
-    }
-    break;
+        //To get actual rs2::stream_profile to open we need to pass an rs2::sensor to `get_required_profile`.
+        std::string requested_stream_name = it.key();
+        if ( ! find_sensor( requested_stream_name, sensor_index ) )
+            throw std::runtime_error( "Could not find sensor to open stream `" + requested_stream_name + "` for" );
```

Comment:
for...

Reply:
Wasn't sure about the English here :-). Will delete the "for"